<!DOCTYPE HTML>
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <title>Genome Browser VCF demo</title>

    <script type="text/javascript" src="../genome-browser.config.js"></script>
    <!-- links for use to build -->
    <link rel="stylesheet" href="../../../node_modules/@fortawesome/fontawesome-free/css/all.min.css">
    <link rel="stylesheet" href="../../../node_modules/qtip2/dist/jquery.qtip.min.css">
    <link rel="stylesheet" href="../../../dist/css/style.css">

    <script type="text/javascript" src="../../../node_modules/underscore/underscore-min.js"></script>
    <script type="text/javascript" src="../../../node_modules/backbone/backbone-min.js"></script>
    <script type="text/javascript" src="../../../node_modules/jquery/dist/jquery.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/qtip2/dist/jquery.qtip.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/urijs/src/URI.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/cookies-js/dist/cookies.js"></script>
    <script type="text/javascript" src="../../../node_modules/crypto-js/crypto-js.js"></script>


    <style type="text/css">
        html {
            width: 100%;
            height: 100%;
            position: relative;
            overflow-x: hidden;
        }
        body {
            margin: 0px;
        }
        #application {
            box-sizing: border-box;
            margin: 0 auto;
            width: 1500px;
            border: 1px solid lightgray;
        }
        #title {
            margin: 20px auto;
            width: 1500px;
            font-size: 35px;
            color: gray;
        }
    </style>
</head>

<body class="ocb">
<div id="title">Genome Browser VCF demo</div>
<div id="application"></div>

<script type="module">
    import { CellBaseClient } from "../../core/clients/cellbase/cellbase-client.js";
    import { OpenCGAClient } from "../../core/clients/opencga/opencga-client.js";
    import GenomeBrowser from "../genome-browser.js";
    import Utils from "../../core/utils.js";
    import Region from "../../core/bioinfo/region.js";
    import FeatureRenderer from "../renderers/feature-renderer.js";
    import SequenceRenderer from "../renderers/sequence-renderer.js";
    import CellBaseAdapter from "../../core/data-adapter/cellbase-adapter.js";
    import FeatureTrack from "../tracks/feature-track.js";
    import GeneTrack from "../tracks/gene-track.js";
    import VariantTrack from "../tracks/variant-track.js";
    /***************************************************************************/
    /***** Setting Defaults Params                                           ***/
    /***************************************************************************/


    let cellBaseClientConfig = {
        host: CELLBASE_HOST,
        version: CELLBASE_VERSION,
        cache: {active: false}
    };
    let cellbaseClient = new CellBaseClient(cellBaseClientConfig);

    let opencgaClientConfig = {
        host: OPENCGA_HOST,
        version: "v2",
        serverVersion: "1.4"
    };

    let opencgaClient = new OpenCGAClient(opencgaClientConfig);
    let AVAILABLE_SPECIES;

//     If you need to login to opencga:
       OPENCGA_USER = "demouser";
       OPENCGA_PASSWORD = "demouser";
       opencgaClient.login(OPENCGA_USER, OPENCGA_PASSWORD)
               .then(function() {
                   console.log("User " + OPENCGA_USER + " successfully logged in in opencga.");
                   opencgaClient = new OpenCGAClient(opencgaClientConfig);
               })
               .catch(function(e) {
                   console.log("Error logging in " + OPENCGA_USER, e);
               });


    getSpecies(function(s) {
        AVAILABLE_SPECIES = s;
        run();
    });

    function getSpecies(callback) {
        cellbaseClient.getMeta("species").then(function(r) {
            let taxonomies = r.response[0].result[0];
            for (let taxonomy in taxonomies) {
                let newSpecies = [];
                for (let i = 0; i < taxonomies[taxonomy].length; i++) {
                    let species = taxonomies[taxonomy][i];
                    for (let j = 0; j < species.assemblies.length; j++) {
                        let s = Utils.clone(species)
                        s.assembly = species.assemblies[j];
                        delete s.assemblies;
                        newSpecies.push(s)
                    }
                }
                taxonomies[taxonomy] = newSpecies;
            }
            callback(taxonomies);
        });
    }

    function run() {
        let species = AVAILABLE_SPECIES.vertebrates[0];
        let region = new Region({chromosome: "13", start: 32996311, end: 33000856}); //initial region
        let genomeBrowser = new GenomeBrowser({
            client: cellbaseClient,
            cellBaseHost: CELLBASE_HOST,
            cellBaseVersion: CELLBASE_VERSION,
            target: 'application',
            width: document.querySelector('#application').getBoundingClientRect().width,
            region: region,
            availableSpecies: AVAILABLE_SPECIES,
            species: species,
            sidePanel: false,
            autoRender: true,
            resizable: true,
            karyotypePanelConfig: {
                collapsed: false,
                collapsible: true
            },
            chromosomePanelConfig: {
                collapsed: false,
                collapsible: true
            },
            navigationBarConfig: {
                componentsConfig: {
                }
            },
            handlers: {
                'region:change': function(e) {
                    console.log(e)
                }
            }
        });

         const tracks = [];

        /***************************************************************************/
        /***** Demo Gene overview Track using FeatureTrack and Cellbase adapter  ***/
        /***************************************************************************/

        let geneOverview = new FeatureTrack({
        title: 'Gene overview',
            minHistogramRegionSize: 20000000,
            maxLabelRegionSize: 10000000,
            height: 80,
            renderer: new FeatureRenderer(FEATURE_TYPES.gene),
            dataAdapter: new CellBaseAdapter(cellbaseClient, "genomic", "region", "gene", {
                exclude: 'transcripts,chunkIds'
            }, {
                chunkSize: 100000
            })
        });
        genomeBrowser.addOverviewTrack(geneOverview);

        /***************************************************************************/
        /***** Demo Sequence Track using FeatureTrack and Cellbase adapter       ***/
        /***************************************************************************/

        const sequence = new FeatureTrack({
            title: 'Sequence',
            height: 20,
            visibleRegionSize: 200,
            renderer: new SequenceRenderer(),
            dataAdapter: new CellBaseAdapter(cellbaseClient, "genomic", "region", "sequence", {}, { chunkSize: 100})
        });
        tracks.push(sequence);

        /**************************************************************************/
        /***** Demo Gene Track using GeneTrack and Cellbase adapter             ***/
        /**************************************************************************/

        const gene = new GeneTrack({
            title: 'Gene',
            minHistogramRegionSize: 20000000,
            maxLabelRegionSize: 10000000,
            minTranscriptRegionSize: 200000,
            height: 120,
            cellbase: {
                "host": CELLBASE_HOST,
                "version": CELLBASE_VERSION,
                "species": "hsapiens"
            }
        });
        tracks.push(gene);

        //The user can change the render and adapter as in the commented example

        /***************************************************************************/
        /*****   Demo Variant Track using VariantTrack and Opencga adapter       ***/
        /***     Using FeatureRender; Visualization of variants individually     ***/
        /***************************************************************************/

        let variant = new VariantTrack({
            title: "Variant",
            closable: true,
            minHistogramRegionSize: 20000000,
            maxLabelRegionSize: 10000000,
            minTranscriptRegionSize: 200000,
            visibleRegionSize: 100000000,
            height: 300,
            opencga: {
//                host: OPENCGA_HOST,
//                version: "v1",
                client: opencgaClient,
//                studies: "reference_grch37:1kG_phase3"
//                 studies: "platinum:illumina_platinum"
                studies: "demo@family:platinum"
//                samples: ["HG00096", "HG00097", "HG00099"]
//                samples: []
            },
//            renderer: new FeatureRenderer(FEATURE_TYPES.variant),
            ////////////////////////////////////////////////////////////////////////////////////////////////////
            //You can define a particular dataAdapter, if not defined, the track will pick up the default one.//
            ///////////////////////////////////////////////////////////////////////////////////////////////////
//            dataAdapter: new OpencgaAdapter(opencgaClient, "analysis/variant", undefined, "query",
//                    {
//                        studies: "reference_grch37:1kG_phase3",
////                      exclude: "studies.files,studies.stats,annotation" //Exclude for samples mode
//                        exclude: "studies,annotation" //Exclude for individual variation mode
//                    }, {
//                        chunkSize: 100000
//                    })
        });
        tracks.push(variant);
        /*******************/

        /***************************************************************************/
        /*****   Demo Variant Track using VariantTrack and Opencga adapter       ***/
        /***     Using FeatureRender; Visualization of variants for a family     ***/
        /***************************************************************************/

        let variant2 = new VariantTrack({
            title: "Variant for family",
            closable: true,
            minHistogramRegionSize: 20000000,
            maxLabelRegionSize: 10000000,
            minTranscriptRegionSize: 200000,
            visibleRegionSize: 100000000,
            height: 300,
            opencga: {
                client: opencgaClient,
                studies: "reference_grch37:1kG_phase3",
                samples: ["HG00096", "HG00097", "HG00099"]
            }
        });
        tracks.push(variant2);
        /*******************/


        genomeBrowser.addTrack(tracks);
        genomeBrowser.draw();
    }
</script>

</body>
</html>
