<!--
  ~ Copyright 2015-2016 OpenCB
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~     http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->

<!DOCTYPE html>

<html lang="en">
<head>

    <meta charset="utf-8">
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <!-- This is configured below in the onLoad() function -->
    <title></title>

    <!-- Favicon -->
    <link rel="icon" href="favicon.ico">

    <!-- Chrome for Android theme color -->
    <meta name="theme-color" content="#000000">

    <!-- Chrome for Android Theme color -->
    <meta name="msapplication-TileColor" content="#3372DF">

    <!-- Add to homescreen for Chrome on Android -->
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="application-name" content="IVA">

    <!-- Add to homescreen for Safari on iOS -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <meta name="apple-mobile-web-app-title" content="IVA">

    <!-- Force Microsoft use latest web tech -->
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <!-- END meta -->

    <!-- build:js[webcomponents/webcomponents-loader.js] -->
    <script type="text/javascript" src="../../../../../../node_modules/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>
    <!-- /build -->

    <!-- build:js[assets/js/vendor.js] -->
    <script type="text/javascript" src="../../../../../../node_modules/jquery/dist/jquery.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/underscore/underscore-min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/backbone/backbone-min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/moment/min/moment.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/highcharts/highcharts.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/qtip2/dist/jquery.qtip.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/urijs/src/URI.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/cookies-js/dist/cookies.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/crypto-js/core.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/crypto-js/sha256.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/bootstrap/dist/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/bootstrap-table/dist/bootstrap-table.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/bootstrap-select/dist/js/bootstrap-select.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/bootstrap-treeview/dist/bootstrap-treeview.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/bootstrap-colorpicker/dist/js/bootstrap-colorpicker.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/bootstrap-validator/dist/validator.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/eonasdan-bootstrap-datetimepicker/build/js/bootstrap-datetimepicker.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/jwt-decode/build/jwt-decode.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/bootstrap-3-typeahead/bootstrap3-typeahead.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/tokenize2/dist/tokenize2.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/@svgdotjs/svg.js/dist/svg.min.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/sweetalert2/dist/sweetalert2.js"></script>
    <script type="text/javascript" src="../../../../../../node_modules/lodash/lodash.min.js"></script>
    <!-- /build -->

    <!-- build:css[assets/css/vendor.css] -->
    <link rel="stylesheet" href="../../../../../../node_modules/bootstrap/dist/css/bootstrap.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/bootstrap-table/dist/bootstrap-table.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/bootstrap-select/dist/css/bootstrap-select.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/bootstrap-treeview/dist/bootstrap-treeview.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/bootstrap-colorpicker/dist/css/bootstrap-colorpicker.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/eonasdan-bootstrap-datetimepicker/build/css/bootstrap-datetimepicker.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/@fortawesome/fontawesome-free/css/all.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/qtip2/dist/jquery.qtip.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/tokenize2/dist/tokenize2.min.css">
    <link rel="stylesheet" href="../../../../../../node_modules/sweetalert2/dist/sweetalert2.css">
    <!-- /build -->
    <link rel="stylesheet" href="../../../../../../../../src/styles/global.css">

    <script src="../../../../../../../../src/conf/config.js"></script>
    <script type="module" src="../variant-interpreter-landing.js"></script>
</head>

<body>
    <script type="module">
            import {LitElement, html} from "/web_modules/lit-element.js";
            import {OpenCGAClient} from "../../../../clients/opencga/opencga-client.js";
            import "../../../../webcomponents/commons/filters/select-field-filter.js";
            import clinicalAnalysis from "./SING-1.js";

            class VariantInterpreterLandingTest extends LitElement {

                constructor() {
                    super();
                }

                createRenderRoot() {
                    return this;
                }

                static get properties() {
                    return {
                    };
                }

                render() {
                    this.clinicalAnalysisId = "SING-1";
                    this.opencgaClient = new OpenCGAClient({
                        host: "http://bioinfo.hpc.cam.ac.uk/opencga-prod",
                        version: "v2",
                        token: "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJhYWx0YW11cmEiLCJhdWQiOiJPcGVuQ0dBIHVzZXJzIiwiaWF0IjoxNjAxNTY3MDk3LCJleHAiOjE2MDE1NzA2OTd9.fDZRqJNW9QZ8C27-UkFIOeggQneziVmHEwWUnxcMXXQ",
                        userId: "aaltamura",
                        cookies: {active: true, prefix: "test"},
                    });
                    this.opencgaSession = {
                        project: {
                            id: "cancer37"
                        },
                        user: {"id":"aaltamura","name":"Antonio Altamura","email":"aaltamura@opencb.org","organization":"","account":{"type":"FULL","creationDate":"","expirationDate":"","authentication":{"id":"internal","application":false}},"internal":{"status":{"name":"READY","date":"20200625130205","description":""}},"quota":{"diskUsage":-1,"cpuUsage":-1,"maxDisk":0,"maxCpu":-1},"projects":[],"configs":{"IVA":{"lastAccess":1601560358862}},"filters":[{"id":"Date","description":"date","resource":"FILE","query":{"study":"demo@family:platinum","creationDate":">=20200903"},"options":{}},{"id":"Recent","description":"","resource":"FILE","query":{"study":"demo@family:platinum","creationDate":">=20200907"},"options":{}}],"attributes":{}},
                        study: {"id":"test","name":"Test","uuid":"f143b23f-0172-0002-0001-b28fde73e5d4","creationDate":"20200626153405","modificationDate":"20200626153405","description":"","size":0,"fqn":"serena@cancer37:test","notification":{},"groups":[{"id":"@members","userIds":["serena","dperez","jcoll","imedina","jtarraga","pfurio","aaltamura","hd10003","ad923","ym255"]},{"id":"@admins","userIds":["imedina","hd10003","ad923","ym255","dperez"]},{"id":"@opencb-team","userIds":["dperez","jcoll","imedina","jtarraga","pfurio","aaltamura"]}],"files":[],"jobs":[],"individuals":[],"families":[],"samples":[],"cohorts":[],"panels":[],"clinicalAnalyses":[],"variableSets":[{"id":"opencga_alignment_stats","name":"opencga_alignment_stats","unique":true,"confidential":false,"description":"OpenCGA alignment stats","variables":[{"id":"readsMappedAndPaired","name":"readsMappedAndPaired","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":10,"dependsOn":"","description":"Reads mapped and paired: paired-end technology bit set + both mates mapped","attributes":{}},{"id":"rawTotalSequences","name":"rawTotalSequences","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":3,"dependsOn":"","description":"Raw total sequences","attributes":{}},{"id":"pairsWithOtherOrientation","name":"pairsWithOtherOrientation","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":38,"dependsOn":"","description":"Pairs with other orientation","attributes":{}},{"id":"inwardOrientedPairs","name":"inwardOrientedPairs","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":36,"dependsOn":"","description":"Inward oriented pairs","attributes":{}},{"id":"filteredSequences","name":"filteredSequences","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":4,"dependsOn":"","description":"Filtered sequences","attributes":{}},{"id":"isSorted","name":"isSorted","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":6,"dependsOn":"","description":"Is sorted","attributes":{}},{"id":"basesMappedCigar","name":"basesMappedCigar","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":22,"dependsOn":"","description":"Bases mapped (cigar): more accurate","attributes":{}},{"id":"maximumLastFragmentLength","name":"maximumLastFragmentLength","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":32,"dependsOn":"","description":"Maximum last fragment length","attributes":{}},{"id":"totalFirstFragmentLength","name":"totalFirstFragmentLength","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":19,"dependsOn":"","description":"Total first fragment length: ignores clipping","attributes":{}},{"id":"insertSizeStandardDeviation","name":"insertSizeStandardDeviation","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":35,"dependsOn":"","description":"Insert size standard deviation","attributes":{}},{"id":"averageQuality","name":"averageQuality","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":33,"dependsOn":"","description":"Average quality","attributes":{}},{"id":"fileId","name":"fileId","category":"","type":"STRING","required":true,"multiValue":false,"allowedValues":[],"rank":1,"dependsOn":"","description":"File ID","attributes":{}},{"id":"insertSizeAverage","name":"insertSizeAverage","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":34,"dependsOn":"","description":"Insert size average","attributes":{}},{"id":"totalLength","name":"totalLength","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":18,"dependsOn":"","description":"Total length: ignores clipping","attributes":{}},{"id":"readsMapped","name":"readsMapped","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":9,"dependsOn":"","description":"Reads mapped","attributes":{}},{"id":"maximumLength","name":"maximumLength","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":30,"dependsOn":"","description":"Maximum length","attributes":{}},{"id":"averageLastFragmentLength","name":"averageLastFragmentLength","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":29,"dependsOn":"","description":"Average last fragment length","attributes":{}},{"id":"readsPaired","name":"readsPaired","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":13,"dependsOn":"","description":"Reads paired: paired-end technology bit set","attributes":{}},{"id":"percentageOfProperlyPairedReads","name":"percentageOfProperlyPairedReads","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":40,"dependsOn":"","description":"Percentage of properly paired reads","attributes":{}},{"id":"basesDuplicated","name":"basesDuplicated","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":24,"dependsOn":"","description":"Bases duplicated","attributes":{}},{"id":"totalLastFragmentLength","name":"totalLastFragmentLength","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":20,"dependsOn":"","description":"Total last fragment length: ignores clipping","attributes":{}},{"id":"sequences","name":"sequences","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":5,"dependsOn":"","description":"Sequences","attributes":{}},{"id":"nonPrimaryAlignments","name":"nonPrimaryAlignments","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":17,"dependsOn":"","description":"Non-primary alignments","attributes":{}},{"id":"readsDuplicated","name":"readsDuplicated","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":14,"dependsOn":"","description":"Reads duplicated: PCR or optical duplicate bit set","attributes":{}},{"id":"errorRate","name":"errorRate","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":26,"dependsOn":"","description":"Error rate: mismatches / bases mapped (cigar)","attributes":{}},{"id":"outwardOrientedPairs","name":"outwardOrientedPairs","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":37,"dependsOn":"","description":"Outward oriented pairs","attributes":{}},{"id":"averageFirstFragmentLength","name":"averageFirstFragmentLength","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":28,"dependsOn":"","description":"Average first fragment length","attributes":{}},{"id":"basesMapped","name":"basesMapped","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":21,"dependsOn":"","description":"Bases mapped: ignores clipping","attributes":{}},{"id":"mismatches","name":"mismatches","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":25,"dependsOn":"","description":"Mismatches: from NM fields","attributes":{}},{"id":"sampleId","name":"sampleId","category":"","type":"STRING","required":true,"multiValue":false,"allowedValues":[],"rank":2,"dependsOn":"","description":"Sample ID","attributes":{}},{"id":"readsMq0","name":"readsMq0","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":15,"dependsOn":"","description":"Reads MQ0: mapped and MQ=0","attributes":{}},{"id":"maximumFirstFragmentLength","name":"maximumFirstFragmentLength","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":31,"dependsOn":"","description":"Maximum first fragment length","attributes":{}},{"id":"pairsOnDifferentChromosomes","name":"pairsOnDifferentChromosomes","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":39,"dependsOn":"","description":"Pairs on different chromosomes","attributes":{}},{"id":"readsProperlyPaired","name":"readsProperlyPaired","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":12,"dependsOn":"","description":"Reads properly paired: proper-pair bit set","attributes":{}},{"id":"readsQcFailed","name":"readsQcFailed","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":16,"dependsOn":"","description":"Reads QC failed","attributes":{}},{"id":"readsUnmapped","name":"readsUnmapped","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":11,"dependsOn":"","description":"Reads unmapped","attributes":{}},{"id":"basesTrimmed","name":"basesTrimmed","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":23,"dependsOn":"","description":"Bases trimmed","attributes":{}},{"id":"firstFragments","name":"firstFragments","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":7,"dependsOn":"","description":"First fragments","attributes":{}},{"id":"averageLength","name":"averageLength","category":"number","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":27,"dependsOn":"","description":"Average length","attributes":{}},{"id":"lastFragments","name":"lastFragments","category":"number","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":8,"dependsOn":"","description":"Last fragments","attributes":{}}],"entities":["FILE"],"release":1,"attributes":{"resource":"variablesets/alignment-stats-variableset.json"}},{"id":"opencga_individual_genetic_checks","name":"opencga_individual_genetic_checks","unique":false,"confidential":false,"description":"OpenCGA individual genetic checks","variables":[{"id":"individualId","name":"individualId","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"dependsOn":"","description":"Individual ID","attributes":{}},{"id":"mendelianErrorsReport","name":"mendelianErrorsReport","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":7,"dependsOn":"","description":"Mendelian errors report","variableSet":[{"id":"numErrors","name":"numErrors","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Number of errors"},{"id":"errorRatio","name":"errorRatio","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Error ratio"},{"id":"chromAggregation","name":"chromAggregation","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Aggregation per chromosome","variableSet":[{"id":"numErrors","name":"numErrors","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Number of errors"},{"id":"codeAggregation","name":"codeAggregation","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Aggregation per error code for that chromosome"},{"id":"chromosome","name":"chromosome","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Chromosome"}]}],"attributes":{}},{"id":"relatednessReport","name":"relatednessReport","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":6,"dependsOn":"","description":"Relatedness report","variableSet":[{"id":"method","name":"method","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Method"},{"id":"scores","name":"scores","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Scores","variableSet":[{"id":"z1","name":"z1","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"Z1"},{"id":"sampleId1","name":"sampleId1","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Sample ID #1"},{"id":"sampleId2","name":"sampleId2","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Sample ID #2"},{"id":"z0","name":"z0","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"Z0"},{"id":"reportedRelation","name":"reportedRelation","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Reported relation"},{"id":"piHat","name":"piHat","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":6,"description":"PI-HAT"},{"id":"z2","name":"z2","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":5,"description":"Z2"}]}],"attributes":{}},{"id":"siblingsIds","name":"siblingsIds","category":"","type":"STRING","required":false,"multiValue":true,"allowedValues":[],"rank":4,"dependsOn":"","description":"Siblings IDs","attributes":{}},{"id":"sampleId","name":"sampleId","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"dependsOn":"","description":"Sample ID","attributes":{}},{"id":"motherId","name":"motherId","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":3,"dependsOn":"","description":"Mother ID","attributes":{}},{"id":"sexReport","name":"sexReport","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":5,"dependsOn":"","description":"Sex report","variableSet":[{"id":"reportedKaryotypicSex","name":"reportedKaryotypicSex","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Reported karyotypic sex"},{"id":"ratioY","name":"ratioY","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"Ratio: Y-chromoxome / autosomic-chromosomes"},{"id":"reportedSex","name":"reportedSex","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Reported sex"},{"id":"inferredKaryotypicSex","name":"inferredKaryotypicSex","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"Inferred karyotypic sex"},{"id":"ratioX","name":"ratioX","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Ratio: X-chromoxome / autosomic-chromosomes"}],"attributes":{}},{"id":"fatherId","name":"fatherId","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":2,"dependsOn":"","description":"Father ID","attributes":{}}],"entities":["INDIVIDUAL"],"release":1,"attributes":{"resource":"variablesets/individual-genetic-checks-variableset.json"}},{"id":"opencga_variant_file_metadata","name":"opencga_variant_file_metadata","unique":true,"confidential":false,"description":"OpenCGA variant file metadata","variables":[{"id":"path","name":"path","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"dependsOn":"","description":"Path to the original file","attributes":{}},{"id":"stats","name":"stats","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":3,"dependsOn":"","description":"Global statistics calculated for this file","variableSet":[{"id":"chromosomeDensity","name":"chromosomeDensity","category":"","type":"MAP_DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":11,"description":"Total density of variants within the chromosome. counts / chromosome.length"},{"id":"qualityAvg","name":"qualityAvg","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":5,"description":"Mean Quality for all the variants with quality"},{"id":"typeCount","name":"typeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":7,"description":"Variants count group by type. e.g. SNP, INDEL, MNP, SNV, ..."},{"id":"chromosomeCount","name":"chromosomeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":10,"description":"Number of variants per chromosome"},{"id":"variantCount","name":"variantCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Number of variants in the variant set"},{"id":"biotypeCount","name":"biotypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":8,"description":"Variants count group by biotype. e.g. protein-coding, miRNA, lncRNA, ..."},{"id":"tiTvRatio","name":"tiTvRatio","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"TiTvRatio = num. transitions / num. transversions"},{"id":"qualityStdDev","name":"qualityStdDev","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":6,"description":"Standard Deviation of the quality"},{"id":"consequenceTypeCount","name":"consequenceTypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":9,"description":"Variants count group by consequence type. e.g. synonymous_variant, missense_variant, stop_lost, ..."},{"id":"filesCount","name":"filesCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"Number of files in the variant set"},{"id":"sampleCount","name":"sampleCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Number of samples in the variant set"},{"id":"filterCount","name":"filterCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"The number of occurrences for each FILTER value in files from this set..  Each file can contain more than one filter value (usually separated by ';').. "}],"attributes":{}},{"id":"id","name":"id","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"dependsOn":"","description":"File id. Will match with the {@link org.opencb.biodata.models.variant.avro.FileEntry#getFileId}","attributes":{}},{"id":"header","name":"header","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":4,"dependsOn":"","description":"The Variant File Header","variableSet":[{"id":"complexLines","name":"complexLines","category":"","type":"OBJECT","required":false,"multiValue":true,"allowedValues":[],"rank":1,"description":"complex lines, e.g. INFO=<ID=NS,Number=1,Type=Integer,Description=\"Number of samples with data\">","variableSet":[{"id":"type","name":"type","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"Type of the values associated with this metadata line.\n        Only present if the metadata line describes data fields, i.e. key == INFO or FORMAT\n        Accepted values:\n          - Integer\n          - Float\n          - String\n          - Character\n          - Flag"},{"id":"genericFields","name":"genericFields","category":"","type":"MAP_STRING","required":false,"multiValue":false,"allowedValues":[],"rank":5,"description":"Other optional fields"},{"id":"description","name":"description","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"The description"},{"id":"id","name":"id","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"ID or Name of the line"},{"id":"number","name":"number","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"Arity of the values associated with this metadata line.\n        Only present if the metadata line describes data fields, i.e. key == INFO or FORMAT\n        Accepted values:\n          - <Integer>: The field has always this number of values.\n          - A: The field has one value per alternate allele.\n          - R: The field has one value for each possible allele, including the reference.\n          - G: The field has one value for each possible genotype\n          - .: The number of possible values varies, is unknown or unbounded."},{"id":"key","name":"key","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Key of group of the Complex Header Line, e.g. INFO, FORMAT, FILTER, ALT, ..."}]},{"id":"version","name":"version","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":""},{"id":"simpleLines","name":"simpleLines","category":"","type":"OBJECT","required":false,"multiValue":true,"allowedValues":[],"rank":2,"description":"simple lines, e.g. fileDate=20090805","variableSet":[{"id":"value","name":"value","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Value"},{"id":"key","name":"key","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Key of group of the Simple Header Line, e.g. source, assembly, pedigreeDB, ..."}]}],"attributes":{}},{"id":"sampleIds","name":"sampleIds","category":"","type":"STRING","required":false,"multiValue":true,"allowedValues":[],"rank":2,"dependsOn":"","description":"Ordered list of sample ids contained in the file","attributes":{}},{"id":"attributes","name":"attributes","category":"","type":"MAP_STRING","required":false,"multiValue":false,"allowedValues":[],"rank":5,"dependsOn":"","description":"Other user defined attributes related with the file","attributes":{}}],"entities":["FILE"],"release":1,"attributes":{"avroClass":"class org.opencb.biodata.models.variant.metadata.VariantFileMetadata","resource":"variablesets/variant-file-metadata-variableset.json"}},{"id":"opencga_cohort_variant_stats","name":"opencga_cohort_variant_stats","unique":true,"confidential":false,"description":"OpenCGA cohort variant stats","variables":[{"id":"chromosomeDensity","name":"chromosomeDensity","category":"","type":"MAP_DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":11,"dependsOn":"","description":"Total density of variants within the chromosome. counts / chromosome.length","attributes":{}},{"id":"qualityAvg","name":"qualityAvg","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":5,"dependsOn":"","description":"Mean Quality for all the variants with quality","attributes":{}},{"id":"typeCount","name":"typeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":7,"dependsOn":"","description":"Variants count group by type. e.g. SNP, INDEL, MNP, SNV, ...","attributes":{}},{"id":"chromosomeCount","name":"chromosomeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":10,"dependsOn":"","description":"Number of variants per chromosome","attributes":{}},{"id":"variantCount","name":"variantCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":0,"dependsOn":"","description":"Number of variants in the variant set","attributes":{}},{"id":"biotypeCount","name":"biotypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"allowedKeys":["3prime_overlapping_ncrna","IG_C_gene","IG_C_pseudogene","IG_D_gene","IG_J_gene","IG_J_pseudogene","IG_V_gene","IG_V_pseudogene","Mt_rRNA","Mt_tRNA","TR_C_gene","TR_D_gene","TR_J_gene","TR_J_pseudogene","TR_V_gene","TR_V_pseudogene","antisense","lincRNA","miRNA","misc_RNA","polymorphic_pseudogene","processed_pseudogene","processed_transcript","protein_coding","pseudogene","rRNA","sense_intronic","sense_overlapping","snRNA","snoRNA","nonsense_mediated_decay","NMD_transcript_variant","unprocessed_pseudogene","transcribed_unprocessed_pseudogene","retained_intron","non_stop_decay","unitary_pseudogene","translated_processed_pseudogene","transcribed_processed_pseudogene","tRNA_pseudogene","snoRNA_pseudogene","snRNA_pseudogene","scRNA_pseudogene","rRNA_pseudogene","misc_RNA_pseudogene","miRNA_pseudogene","non_coding","ambiguous_orf","known_ncrna","retrotransposed","transcribed_unitary_pseudogene","translated_unprocessed_pseudogene","LRG_gene","intergenic_variant"],"rank":8,"dependsOn":"","description":"Variants count group by biotype. e.g. protein-coding, miRNA, lncRNA, ...","attributes":{}},{"id":"tiTvRatio","name":"tiTvRatio","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":4,"dependsOn":"","description":"TiTvRatio = num. transitions / num. transversions","attributes":{}},{"id":"qualityStdDev","name":"qualityStdDev","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":6,"dependsOn":"","description":"Standard Deviation of the quality","attributes":{}},{"id":"filesCount","name":"filesCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":3,"dependsOn":"","description":"Number of files in the variant set","attributes":{}},{"id":"consequenceTypeCount","name":"consequenceTypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"allowedKeys":["start_retained_variant","upstream_variant","3_prime_UTR_variant","splice_acceptor_variant","transcript_amplification","upstream_gene_variant","RNA_polymerase_promoter","non_coding_transcript_exon_variant","non_coding_transcript_variant","inframe_variant","transcript_ablation","splice_donor_variant","synonymous_variant","feature_elongation","feature_truncation","miRNA_target_site","exon_variant","downstream_gene_variant","stop_retained_variant","TF_binding_site_variant","initiator_codon_variant","coding_sequence_variant","protein_altering_variant","intergenic_variant","terminator_codon_variant","frameshift_variant","DNAseI_hypersensitive_site","feature_variant","2KB_downstream_variant","intron_variant","splice_region_variant","5_prime_UTR_variant","SNP","stop_gained","regulatory_region_amplification","2KB_upstream_variant","miRNA","lincRNA","start_lost","SNV","CpG_island","downstream_variant","NMD_transcript_variant","2KB_downstream_gene_variant","TFBS_amplification","missense_variant","regulatory_region_ablation","mature_miRNA_variant","stop_lost","structural_variant","regulatory_region_variant","TFBS_ablation","copy_number_change","2KB_upstream_gene_variant","polypeptide_variation_site","inframe_deletion","inframe_insertion","incomplete_terminal_codon_variant"],"rank":9,"dependsOn":"","description":"Variants count group by consequence type. e.g. synonymous_variant, missense_variant, stop_lost, ...","attributes":{}},{"id":"sampleCount","name":"sampleCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":1,"dependsOn":"","description":"Number of samples in the variant set","attributes":{}},{"id":"filterCount","name":"filterCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"dependsOn":"","description":"The number of occurrences for each FILTER value in files from this set..  Each file can contain more than one filter value (usually separated by ';').. ","attributes":{}}],"entities":["COHORT"],"release":1,"attributes":{"avroClass":"class org.opencb.biodata.models.variant.metadata.VariantSetStats","resource":"variablesets/cohort-variant-stats-variableset.json"}},{"id":"opencga_sample_variant_stats","name":"opencga_sample_variant_stats","unique":true,"confidential":false,"description":"OpenCGA sample variant stats","variables":[{"id":"indelLengthCount","name":"indelLengthCount","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":5,"dependsOn":"","description":"Indel length grouped in ranges *","variableSet":[{"id":"lt20","name":"lt20","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":""},{"id":"lt5","name":"lt5","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":""},{"id":"lt10","name":"lt10","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":""},{"id":"lt15","name":"lt15","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":""},{"id":"gte20","name":"gte20","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":""}],"attributes":{}},{"id":"consequenceTypeCount","name":"consequenceTypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"allowedKeys":["start_retained_variant","upstream_variant","3_prime_UTR_variant","splice_acceptor_variant","transcript_amplification","upstream_gene_variant","RNA_polymerase_promoter","non_coding_transcript_exon_variant","non_coding_transcript_variant","inframe_variant","transcript_ablation","splice_donor_variant","synonymous_variant","feature_elongation","feature_truncation","miRNA_target_site","exon_variant","downstream_gene_variant","stop_retained_variant","TF_binding_site_variant","initiator_codon_variant","coding_sequence_variant","protein_altering_variant","intergenic_variant","terminator_codon_variant","frameshift_variant","DNAseI_hypersensitive_site","feature_variant","2KB_downstream_variant","intron_variant","splice_region_variant","5_prime_UTR_variant","SNP","stop_gained","regulatory_region_amplification","2KB_upstream_variant","miRNA","lincRNA","start_lost","SNV","CpG_island","downstream_variant","NMD_transcript_variant","2KB_downstream_gene_variant","TFBS_amplification","missense_variant","regulatory_region_ablation","mature_miRNA_variant","stop_lost","structural_variant","regulatory_region_variant","TFBS_ablation","copy_number_change","2KB_upstream_gene_variant","polypeptide_variation_site","inframe_deletion","inframe_insertion","incomplete_terminal_codon_variant"],"rank":12,"dependsOn":"","description":"Variants count group by consequence type. e.g. missense_variant, synonymous_variant, stop_lost, ....  Each counter is increased at most one per variant. If multiple overlapping transcripts have the same consequence type, it will count as one.","attributes":{}},{"id":"genotypeCount","name":"genotypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":4,"dependsOn":"","description":"Number of variants per genotype. Only counts genotypes with the main allele. Phase is ignored. *","attributes":{}},{"id":"id","name":"id","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"dependsOn":"","description":"Sample identifier *","attributes":{}},{"id":"qualityStdDev","name":"qualityStdDev","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":9,"dependsOn":"","description":"Standard Deviation of the quality","attributes":{}},{"id":"filterCount","name":"filterCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":6,"dependsOn":"","description":"The number of occurrences for each FILTER value in files from this set..  Each file can contain more than one filter value (usually separated by ';').. ","attributes":{}},{"id":"typeCount","name":"typeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":3,"dependsOn":"","description":"Variants count group by type. e.g. SNP, INDEL, MNP, SNV, ...","attributes":{}},{"id":"heterozygosityRate","name":"heterozygosityRate","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":10,"dependsOn":"","description":"Heterozygosity rate as defined by PLINK: (N–O)/N. .  N is the number of non-missing genotypes.  O is the observed number of homozygous genotypes for a given individual. ","attributes":{}},{"id":"mendelianErrorCount","name":"mendelianErrorCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":11,"dependsOn":"","description":"Number of mendelian errors grouped by PLINK error codes. *","attributes":{}},{"id":"variantCount","name":"variantCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":1,"dependsOn":"","description":"Number of variants where the sample has the main allele (i.e. 0/1, 1/1, ./1, 1/2, ...)","attributes":{}},{"id":"biotypeCount","name":"biotypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"allowedKeys":["3prime_overlapping_ncrna","IG_C_gene","IG_C_pseudogene","IG_D_gene","IG_J_gene","IG_J_pseudogene","IG_V_gene","IG_V_pseudogene","Mt_rRNA","Mt_tRNA","TR_C_gene","TR_D_gene","TR_J_gene","TR_J_pseudogene","TR_V_gene","TR_V_pseudogene","antisense","lincRNA","miRNA","misc_RNA","polymorphic_pseudogene","processed_pseudogene","processed_transcript","protein_coding","pseudogene","rRNA","sense_intronic","sense_overlapping","snRNA","snoRNA","nonsense_mediated_decay","NMD_transcript_variant","unprocessed_pseudogene","transcribed_unprocessed_pseudogene","retained_intron","non_stop_decay","unitary_pseudogene","translated_processed_pseudogene","transcribed_processed_pseudogene","tRNA_pseudogene","snoRNA_pseudogene","snRNA_pseudogene","scRNA_pseudogene","rRNA_pseudogene","misc_RNA_pseudogene","miRNA_pseudogene","non_coding","ambiguous_orf","known_ncrna","retrotransposed","transcribed_unitary_pseudogene","translated_unprocessed_pseudogene","LRG_gene","intergenic_variant"],"rank":13,"dependsOn":"","description":"Variants count group by biotype. e.g. protein-coding, miRNA, lncRNA, ....  Each counter is increased at most one per variant. If multiple overlapping genes have the same biotypes, it will count as one.","attributes":{}},{"id":"chromosomeCount","name":"chromosomeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"dependsOn":"","description":"Number of variants per chromosome *","attributes":{}},{"id":"qualityAvg","name":"qualityAvg","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":8,"dependsOn":"","description":"Mean Quality for all the variants with quality","attributes":{}},{"id":"tiTvRatio","name":"tiTvRatio","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":7,"dependsOn":"","description":"TiTvRatio = num. transitions / num. transversions","attributes":{}}],"entities":["SAMPLE"],"release":1,"attributes":{"avroClass":"class org.opencb.biodata.models.variant.metadata.SampleVariantStats","resource":"variablesets/sample-variant-stats-variableset.json"}},{"id":"opencga_file_variant_stats","name":"opencga_file_variant_stats","unique":true,"confidential":false,"description":"OpenCGA file variant stats","variables":[{"id":"chromosomeDensity","name":"chromosomeDensity","category":"","type":"MAP_DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":11,"dependsOn":"","description":"Total density of variants within the chromosome. counts / chromosome.length","attributes":{}},{"id":"typeCount","name":"typeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":7,"dependsOn":"","description":"Variants count group by type. e.g. SNP, INDEL, MNP, SNV, ...","attributes":{}},{"id":"qualityAvg","name":"qualityAvg","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":5,"dependsOn":"","description":"Mean Quality for all the variants with quality","attributes":{}},{"id":"chromosomeCount","name":"chromosomeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":10,"dependsOn":"","description":"Number of variants per chromosome","attributes":{}},{"id":"variantCount","name":"variantCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":0,"dependsOn":"","description":"Number of variants in the variant set","attributes":{}},{"id":"biotypeCount","name":"biotypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"allowedKeys":["3prime_overlapping_ncrna","IG_C_gene","IG_C_pseudogene","IG_D_gene","IG_J_gene","IG_J_pseudogene","IG_V_gene","IG_V_pseudogene","Mt_rRNA","Mt_tRNA","TR_C_gene","TR_D_gene","TR_J_gene","TR_J_pseudogene","TR_V_gene","TR_V_pseudogene","antisense","lincRNA","miRNA","misc_RNA","polymorphic_pseudogene","processed_pseudogene","processed_transcript","protein_coding","pseudogene","rRNA","sense_intronic","sense_overlapping","snRNA","snoRNA","nonsense_mediated_decay","NMD_transcript_variant","unprocessed_pseudogene","transcribed_unprocessed_pseudogene","retained_intron","non_stop_decay","unitary_pseudogene","translated_processed_pseudogene","transcribed_processed_pseudogene","tRNA_pseudogene","snoRNA_pseudogene","snRNA_pseudogene","scRNA_pseudogene","rRNA_pseudogene","misc_RNA_pseudogene","miRNA_pseudogene","non_coding","ambiguous_orf","known_ncrna","retrotransposed","transcribed_unitary_pseudogene","translated_unprocessed_pseudogene","LRG_gene","intergenic_variant"],"rank":8,"dependsOn":"","description":"Variants count group by biotype. e.g. protein-coding, miRNA, lncRNA, ...","attributes":{}},{"id":"tiTvRatio","name":"tiTvRatio","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":4,"dependsOn":"","description":"TiTvRatio = num. transitions / num. transversions","attributes":{}},{"id":"qualityStdDev","name":"qualityStdDev","category":"","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":6,"dependsOn":"","description":"Standard Deviation of the quality","attributes":{}},{"id":"filesCount","name":"filesCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":3,"dependsOn":"","description":"Number of files in the variant set","attributes":{}},{"id":"consequenceTypeCount","name":"consequenceTypeCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"allowedKeys":["start_retained_variant","upstream_variant","3_prime_UTR_variant","splice_acceptor_variant","transcript_amplification","upstream_gene_variant","RNA_polymerase_promoter","non_coding_transcript_exon_variant","non_coding_transcript_variant","inframe_variant","transcript_ablation","splice_donor_variant","synonymous_variant","feature_elongation","feature_truncation","miRNA_target_site","exon_variant","downstream_gene_variant","stop_retained_variant","TF_binding_site_variant","initiator_codon_variant","coding_sequence_variant","protein_altering_variant","intergenic_variant","terminator_codon_variant","frameshift_variant","DNAseI_hypersensitive_site","feature_variant","2KB_downstream_variant","intron_variant","splice_region_variant","5_prime_UTR_variant","SNP","stop_gained","regulatory_region_amplification","2KB_upstream_variant","miRNA","lincRNA","start_lost","SNV","CpG_island","downstream_variant","NMD_transcript_variant","2KB_downstream_gene_variant","TFBS_amplification","missense_variant","regulatory_region_ablation","mature_miRNA_variant","stop_lost","structural_variant","regulatory_region_variant","TFBS_ablation","copy_number_change","2KB_upstream_gene_variant","polypeptide_variation_site","inframe_deletion","inframe_insertion","incomplete_terminal_codon_variant"],"rank":9,"dependsOn":"","description":"Variants count group by consequence type. e.g. synonymous_variant, missense_variant, stop_lost, ...","attributes":{}},{"id":"sampleCount","name":"sampleCount","category":"","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":1,"dependsOn":"","description":"Number of samples in the variant set","attributes":{}},{"id":"filterCount","name":"filterCount","category":"","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"dependsOn":"","description":"The number of occurrences for each FILTER value in files from this set..  Each file can contain more than one filter value (usually separated by ';').. ","attributes":{}}],"entities":["FILE"],"release":1,"attributes":{"avroClass":"class org.opencb.biodata.models.variant.metadata.VariantSetStats","resource":"variablesets/file-variant-stats-variableset.json"}},{"id":"opencga_sample_qc","name":"opencga_sample_qc","unique":true,"confidential":false,"description":"OpenCGA variable set for sample quality control (QC)","variables":[{"id":"relatednessReport","name":"Relatedness report","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":6,"dependsOn":"","description":"Relatedness report","variableSet":[{"id":"scores","name":"Scores","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Scores","variableSet":[{"id":"sampleId1","name":"ID for sample #1","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Sample ID #1"},{"id":"z2","name":"Z2 value","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":5,"description":"Z2 value"},{"id":"sampleId2","name":"ID for sample #2","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Sample ID #2"},{"id":"piHat","name":"PI-HAT value","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":6,"description":"PI-HAT value"},{"id":"z1","name":"Z1 value","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"Z1 value"},{"id":"z0","name":"Z0 value","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"Z0 value"},{"id":"reportedRelation","name":"Reported relation","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Reported relation"}]},{"id":"method","name":"Method","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Method"}],"attributes":{}},{"id":"siblingsIds","name":"List of siblings IDs","category":"","type":"STRING","required":false,"multiValue":true,"allowedValues":[],"rank":4,"dependsOn":"","description":"List of siblings IDs","attributes":{}},{"id":"sampleId","name":"Sample ID","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"dependsOn":"","description":"Sample ID","attributes":{}},{"id":"hsMetricsReport","name":"Hs metrics report","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":10,"dependsOn":"","description":"Hs metrics report (from the picard/CollecHsMetrics command)","variableSet":[{"id":"minTargetCoverage","name":"Min target coverage","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":23,"description":"The minimum coverage of targets"},{"id":"totalReads","name":"Total reads","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":18,"description":"The total number of reads in the SAM or BAM file examine"},{"id":"pfBases","name":"Pf bases","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":12,"description":"Pf bases"},{"id":"hsPenalty","name":"Hybrid selection penalty to 10x, 20x, 30x, 40x, 50x, 100x","type":"DOUBLE","required":false,"multiValue":true,"allowedValues":[],"rank":32,"description":"The hybrid selection penalty incurred to get 80% of target bases to 10X, 20x, 30x, 40x, 50x and 100x"},{"id":"hetSnpSensitivity","name":"Het snp sensitivity","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":7,"description":"The theoretical HET SNP sensitivity"},{"id":"zeroCvgTargetsPct","name":"Zero cvg targets pct","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":26,"description":"The fraction of targets that did not reach coverage=1 over any base"},{"id":"pfUqBasesAligned","name":"Pf uq bases aligned","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":16,"description":"The number of bases in the PF aligned reads that are mapped to a reference base. Accounts for clipping and gaps"},{"id":"zeroCvgTargetsPct","name":"Zero cvg targets pct","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":27,"description":"The fraction of targets that did not reach coverage=1 over any base"},{"id":"onBaitVsSelected","name":"On bait vs selected","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":24,"description":"The percentage of on+near bait bases that are on as opposed to near"},{"id":"offBaitBases","name":"Off bait bases","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":9,"description":"The number of PF aligned bases that mapped to neither on or near a bait"},{"id":"medianTargetCoverage","name":"Median target coverage","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":22,"description":"The median coverage of targets"},{"id":"fold80BasePenalty","name":"Fold 80 base penalty","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"The fold over-coverage necessary to raise 80% of bases in \"non-zero-cvg\" targets to the mean coverage level in those targets"},{"id":"pfBasesAligned","name":"Pf bases aligned","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":13,"description":"The number of PF unique bases that are aligned with mapping score > 0 to the reference genome"},{"id":"meanTargetCoverage","name":"Mean target coverage","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":21,"description":"The mean coverage of targets"},{"id":"nearBaitBases","name":"Near bait bases","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":8,"description":"The number of PF aligned bases that mapped to within a fixed interval of a baited region, but not on a baited region"},{"id":"targetTerritory","name":"Target territory","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":25,"description":"The unique number of target bases in the experiment where target is usually exons etc."},{"id":"pctTargetBases","name":"Pct target bases 1x, 2x, 10x, 20x, 30x, 40x, 50x, 100x","type":"DOUBLE","required":false,"multiValue":true,"allowedValues":[],"rank":31,"description":"The fraction of all target bases achieving 1x, 2x, 10x, 20x, 30x, 40x, 50x and 100x"},{"id":"pfReads","name":"Pf reads","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":14,"description":"The number of reads that pass the vendor's filter"},{"id":"foldEnrichment","name":"Fold enrichment","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"The fold by which the baited region has been amplified above genomic background"},{"id":"pfUqReadsAligned","name":"Pf uq reads aligned","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":17,"description":"The number of PF unique reads that are aligned with mapping score > 0 to the reference genome"},{"id":"gcDropout","name":"Gc dropout","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":5,"description":"A measure of how undercovered >= 50% GC regions are relative to the mean"},{"id":"meanBaitCoverage","name":"Mean bait coverage","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":20,"description":"The mean coverage of all baits in the experiment"},{"id":"onBaitBases","name":"On bait bases","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":10,"description":"The number of PF aligned bases that mapped to a baited region of the genome"},{"id":"Genome size","name":"Genome size","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":30,"description":"The number of bases in the reference genome used for alignment"},{"id":"pfUniqueReads","name":"Pf unique reads","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":15,"description":"The number of PF reads that are not marked as duplicates"},{"id":"hsLibrarySize","name":"Hs library size","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":28,"description":"The estimated number of unique molecules in the selected part of the library"},{"id":"baitTerritory","name":"Bait territory","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"The number of bases which have one or more baits on top of them"},{"id":"hetSnpQ","name":"Het snp q","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":6,"description":"The Phred Scaled Q Score of the theoretical HET SNP sensitivity"},{"id":"atDropout","name":"A measure of how undercovered <= 50% GC regions are relative to the mean","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"A measure of how undercovered <= 50% GC regions are relative to the mean"},{"id":"baitDesignEfficiency","name":"Bait design efficiency","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Target terrirtoy / bait territory. 1 == perfectly efficient, 0.5 = half of baited bases are not target"},{"id":"baitSet","name":"Bait set","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":29,"description":"The name of the bait set used in the hybrid selection"},{"id":"onTargetBases","name":"On target bases","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":11,"description":"The number of PF aligned bases that mapped to a targeted region of the genome"},{"id":"maxTargetCoverage","name":"Max target coverage","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":19,"description":"Max target coverage"}],"attributes":{}},{"id":"motherId","name":"Mother ID","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":3,"dependsOn":"","description":"Mother ID","attributes":{}},{"id":"individualId","name":"Individual ID","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"dependsOn":"","description":"Individual ID","attributes":{}},{"id":"mendelianErrorsReport","name":"Mendelian errors report","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":7,"dependsOn":"","description":"Mendelian errors report","variableSet":[{"id":"sampleId","name":"Sample ID","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":7,"description":"Sample ID"},{"id":"numErrors","name":"Total number of errors","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Total number of errors"},{"id":"errorRatio","name":"Error ratio","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Error ratio"},{"id":"chromAggregation","name":"Aggregation per chromosome","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Aggregation per chromosome","variableSet":[{"id":"numErrors","name":"Total number of errors","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Total number of errors"},{"id":"codeAggregation","name":"Aggregation per error code","type":"MAP_INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Aggregation per error code for that chromosome"},{"id":"chromosome","name":"chromosome","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Chromosome"}]}],"attributes":{}},{"id":"sexReport","name":"Sex report","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":5,"dependsOn":"","description":"Sex report","variableSet":[{"id":"sampleId","name":"Sample ID","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Sample ID"},{"id":"reportedSex","name":"Reported sex","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Reported sex"},{"id":"inferredKaryotypicSex","name":"Inferred karyotypic sex","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":5,"description":"Inferred karyotypic sex"},{"id":"ratioX","name":"Ratio for chromosome X","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"Ratio: X-chromoxome / autosomic-chromosomes"},{"id":"reportedKaryotypicSex","name":"Reported karyotypic sex","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Reported karyotypic sex"},{"id":"ratioY","name":"Ratio for chromosome Y","type":"DOUBLE","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"Ratio: Y-chromoxome / autosomic-chromosomes"}],"attributes":{}},{"id":"flagStatsReport","name":"Flag stats report","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":9,"dependsOn":"","description":"Flag stats report (from the samtools/flagstat command)","variableSet":[{"id":"properlyPaired","name":"Total number of reads which are properly paired","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":6,"description":"Total number of reads which are properly paired (both 0x1 and 0x2 bits set and 0x4 bit not set)"},{"id":"totalReads","name":"Total number of reads","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":0,"description":"Total number of reads"},{"id":"mapped","name":"Total number of reads which are mapped","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":2,"description":"Total number of reads which are mapped (0x4 bit not set)"},{"id":"diffChrMapQ5","name":"Total number of reads with mate mapped to a different chromosome and with mapping quality at least 5","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":10,"description":"Total number of reads with mate mapped to a different chromosome and with mapping quality at least 5 (0x1 bit set and neither 0x4 nor 0x8 bits set and MRNM not equal to RNAME and MAPQ >= 5)"},{"id":"totalQcPassed","name":"Total QC passed","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":1,"description":"Total number of reads marked as not passing quality controls"},{"id":"mateMappedToDiffChr","name":"Total number of reads with mate mapped to a different chromosome","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":9,"description":"Total number of reads with mate mapped to a different chromosome (0x1 bit set and neither 0x4 nor 0x8 bits set and MRNM not equal to RNAME)"},{"id":"secondaryAlignments","name":"Total number of reads which are secondary","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":3,"description":"Total number of reads which are secondary (0x100 bit set)"},{"id":"pairedInSequencing","name":"Total number of reads which are paired in sequencing","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":5,"description":"Total number of reads which are paired in sequencing (0x1 bit set)"},{"id":"selfAndMateMapped","name":"Total number of reads with itself and mate mapped","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":7,"description":"Total number of reads with itself and mate mapped (0x1 bit set and neither 0x4 nor 0x8 bits set)"},{"id":"singletons","name":"Total number of reads which are singletons","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":8,"description":"Total number of reads which are singletons (both 0x1 and 0x8 bits set and bit 0x4 not set)"},{"id":"duplicates","name":"Total number of reads which are duplicates","type":"INTEGER","required":false,"multiValue":false,"allowedValues":[],"rank":4,"description":"Total number of reads which are duplicates (0x400 bit set)"}],"attributes":{}},{"id":"fatherId","name":"Father ID","category":"","type":"STRING","required":false,"multiValue":false,"allowedValues":[],"rank":2,"dependsOn":"","description":"Father ID","attributes":{}},{"id":"fastQcReport","name":"FastQC report","category":"","type":"OBJECT","required":false,"multiValue":false,"allowedValues":[],"rank":8,"dependsOn":"","description":"FastQC report (from the FastQC tool)","variableSet":[],"attributes":{}}],"entities":["SAMPLE"],"release":1,"attributes":{"resource":"variablesets/sample-qc-variableset.json"}}],"permissionRules":{},"uri":"file:///mnt/data/opencga-prod/sessions/users/serena/projects/202/203/","release":1,"status":{"name":"","description":"","date":""},"internal":{"status":{"name":"READY","date":"20200626153405","description":""}},"attributes":{},"alias":"test","acl":{"@opencb-team":["VIEW_FILES","VIEW_FILE_ANNOTATIONS","WRITE_INDIVIDUALS","VIEW_COHORTS","VIEW_FAMILY_ANNOTATIONS","WRITE_FAMILIES","VIEW_FILE_HEADER","VIEW_FILE_CONTENT","VIEW_INDIVIDUALS","VIEW_AGGREGATED_VARIANTS","VIEW_COHORT_ANNOTATIONS","WRITE_SAMPLES","WRITE_CLINICAL_ANALYSIS","DELETE_JOBS","EXECUTE_JOBS","DOWNLOAD_FILES","VIEW_INDIVIDUAL_ANNOTATIONS","VIEW_PANELS","VIEW_FAMILIES","VIEW_JOBS","WRITE_SAMPLE_ANNOTATIONS","WRITE_JOBS","VIEW_SAMPLE_VARIANTS","WRITE_FAMILY_ANNOTATIONS","VIEW_SAMPLES","WRITE_INDIVIDUAL_ANNOTATIONS","VIEW_SAMPLE_ANNOTATIONS","VIEW_CLINICAL_ANALYSIS"]}},
                        opencgaClient: this.opencgaClient
                    };

                    return html`
                        <variant-interpreter-landing .opencgaSession="${this.opencgaSession}"
                                                .clinicalAnalysis="${clinicalAnalysis}">
                        </variant-interpreter-landing>
                    `;
                }
            }

            customElements.define("variant-interpreter-landing-test", VariantInterpreterLandingTest);
    </script>

    <variant-interpreter-landing-test></variant-interpreter-landing-test>

</body>
</html>
